src/signal_processing_algorithms/energy_statistics/energy_statistics.py

```diff
@@ -0,0 +1,408 @@
+    """
+    Return valid form of input.
+
+    :param series: A distribution.
+    :return: Return a valid input.
+    """
+    if series is None or len(series) == 0:
+        raise ValueError("Distribution cannot be empty.")
+
+    return _convert(series)
+
+
+def _get_distance_matrix(series: np.ndarray, use_c_if_possible: bool = True) -> np.ndarray:
+    """
+    Return the matrix of pairwise distances within the series.
+
+    :param series: An m x n array of m observations for n variables.
+    :return: An m x m array where (i,j)th value is the distance between the observation
+    at i-th row of series and j-th row of series.
+    """
+    if use_c_if_possible and C_EXTENSION_LOADED and series.shape[1] == 1:
+        return calculate_distance_matrix(series.flatten())
+    else:
+        return np.linalg.norm(series[:, np.newaxis] - series, axis=2)
+
+
+def _calculate_stats(x: float, y: float, xy: float, n: int, m: int) -> Tuple[float, float, float]:
+    """
+    Calculate the stats from the terms and coefficients.
+
+    :param xy: The sum of distances between sample distributions X, Y.
+    :param x: The sum of distances within sample distribution X.
+    :param y: The sum of distances within sample distribution Y.
+    :param n: The length of sample distribution X.
+    :param m: The length of sample distribution Y.
+
+    :return: The q value generated from the terms.
+    """
+    xy_avg = xy / (n * m) if n > 0 and m > 0 else 0
+    x_avg = x / (n ** 2) if n > 0 else 0
+    y_avg = y / (m ** 2) if m > 0 else 0
+
+    # E-statistic
+    e = 2 * xy_avg - x_avg - y_avg
+    # Test statistic
+    t = (n * m / (n + m)) * e
+    # E-coefficient of inhomogeneity
+    h = (e / (2 * xy_avg)) if xy_avg > 0 else 0.0
+    return e, t, h
+
+
+def _calculate_t_stats(distance_matrix: np.ndarray, use_c_if_possible: bool = True) -> np.ndarray:
+    """
+    Find t-statistic values given a distance matrix .
+
+    :param distance_matrix: The distance matrix.
+    :return: The t-statistic values.
+    """
+    if use_c_if_possible and C_EXTENSION_LOADED:
+        return calculate_t_values(distance_matrix)
+
+    statistics = np.zeros(len(distance_matrix), dtype=np.float64)
+
+    # We will partition our signal into:
+    # X = {Xi; 0 <= i < tau}
+    # Y = {Yj; tau <= j < len(signal) }
+    # and look for argmax(tau)Q(tau)
+
+    # sum |Xi - Yj| for i < tau <= j
+    xy = 0
+    # sum |Xi - Xj| for i < j < tau
+    x = 0
+    # sum |Yi - Yj| for tau <= i < j
+    y = 0
+
+    for row in range(0, len(distance_matrix)):
+        y += np.sum(distance_matrix[row, row:])
+
+    for tau in range(0, len(distance_matrix)):
+        # Since the distance matrix is symmetric,
+        # we can consider the upper or lower diagonal triangle and double the sum for
+        # calculating the pairwise distances within x as well as y.
+        statistics[tau] = _calculate_stats(2 * x, 2 * y, xy, tau, len(distance_matrix) - tau)[1]
```

Comment:
Same comment, why doubling within distances, but not between? Seems like if we are only taking the sum of the upper triangle, we should double all terms to equal the sum of the whole matrix due to symmetry.